### PR TITLE
Implement signed struct.

### LINKED ISSUE #275: Implement generic signed_struct
In GitLab by @damip

Add this file to models: [signed_struct.rs](/uploads/81c2c1a5cd984836d906d1c40bf8e347/signed_struct.rs)

Then BlockHeader can simply be defined as:

`pub type BlockHeader = SignedStruct<BlockHeaderContent>;`

without any impl

In the same way, an operation will be:

`pub type Operation = SignedStruct<OperationContent>;`

### PR DESCRIPTION
**Pull request by AureliaDolo**
Fixed #275

### CODE REVIEW

**Comment by damip** [comment]
let's keep in sync with Greg's serialization  and data structure refactoring !

**Comment by AureliaDolo** [comment]
I did it for the block headers y'all can take a look to tell if it's too much :)

**Review by AurelienFT — COMMENTED** [review]
As a general note, everywhere you `unwrap()` the error from the `new_signed` which can return an error if `compute_id()` does, which returns an error if `to_bytes_compact()` does. In some implementations of `to_bytes_compact()` we return error in case that (I assume) can really happens such as here : https://github.com/massalabs/massa/blob/main/massa-consensus-exports/src/boostrapable_graph.rs#L31 .

Is that cases that can't happens ? Or just we don't care ?

File: massa-pool/src/endorsement_pool.rs [hunk 1]
```diff
@@ -44,7 +45,7 @@ impl EndorsementPool {
         target_slot: Slot,
         parent: BlockId,
         creators: Vec<Address>,
-    ) -> Result<Vec<(EndorsementId, Endorsement)>, PoolError> {
+    ) -> Result<Vec<(EndorsementId, Signed<Endorsement, EndorsementId>)>, PoolError> {
```
[the remark this answers is not in the document]
Comment (AureliaDolo) [reply massa-pool/src/endorsement_pool.rs:48]: `type SignedEndorsement = Signed<Endorsement, EndorsementId>` that looks like the better option :)

File: massa-models/src/signed.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        writeln!(f, "{}", self.content)?;
+        Ok(())
+    }
+}
+pub trait Signable<U>
+where
+    U: Id,
+    Self: SerializeCompact,
+{
+    fn get_signature_message(&self) -> Result<Hash, ModelsError>;
```
[the remark this answers is not in the document]
Comment (AureliaDolo) [reply massa-models/src/signed.rs:39]: You're right for the default impl stuff. 

But as for using get signature message in compute id seems weird from a semantic point of view, in my opinion.

**Review by AurelienFT — DISMISSED** [review]
If the unwrap that I spoke about is not a problem the rest is cool !

**Comment by AureliaDolo** [comment]
> As a general note, everywhere you `unwrap()` the error from the `new_signed` which can return an error if `compute_id()` does, which returns an error if `to_bytes_compact()` does. In some implementations of `to_bytes_compact()` we return error in case that (I assume) can really happens such as here : https://github.com/massalabs/massa/blob/main/massa-consensus-exports/src/boostrapable_graph.rs#L31 .
> 
> Is that cases that can't happens ? Or just we don't care ?

In tests we don't care about unwraps. Did you see some unwrap anywhere else ?

**Comment by AurelienFT** [comment]
> In tests we don't care about unwraps. Did you see some unwrap anywhere else ?

It's on test. It's ok then :)

**Review by adrien-zinger — DISMISSED** [review]

**Comment by AureliaDolo** [comment]
bors merge

**Review by AurelienFT — APPROVED** [review]

**Review by adrien-zinger — APPROVED** [review]

**Comment by AureliaDolo** [comment]
bors merge

**Pull request merged**